Code/Framework/AzFramework/AzFramework/Spawnable/SpawnableAssetRef.cpp

```diff
@@ -0,0 +1,118 @@
+                ->Field("Asset", &SpawnableAssetRef::m_asset);
+
+            serializeContext->RegisterGenericType<AZStd::vector<SpawnableAssetRef>>();
+            serializeContext->RegisterGenericType<AZStd::unordered_map<AZStd::string, SpawnableAssetRef>>();
+            serializeContext->RegisterGenericType<AZStd::unordered_map<double, SpawnableAssetRef>>(); // required to support Map<Number, SpawnableAssetRef> in Script Canvas
+
+            if (AZ::EditContext* editContext = serializeContext->GetEditContext())
+            {
+                editContext
+                    ->Class<SpawnableAssetRef>("SpawnableAssetRef", "A wrapper around spawnable asset to be used as a variable in Script Canvas.")
+                    // m_asset
+                    ->DataElement(AZ::Edit::UIHandlers::Default, &SpawnableAssetRef::m_asset, "asset", "")
+                    ->Attribute(AZ::Edit::Attributes::ShowProductAssetFileName, false)
+                    ->Attribute(AZ::Edit::Attributes::HideProductFilesInAssetPicker, true)
+                    ->Attribute(AZ::Edit::Attributes::AssetPickerTitle, "Spawnable Asset")
+                    ->Attribute(AZ::Edit::Attributes::ChangeNotify, &SpawnableAssetRef::OnSpawnAssetChanged);
+            }
+        }
+
+        if (auto behaviorContext = azrtti_cast<AZ::BehaviorContext*>(context))
+        {
+            behaviorContext
+                ->Class<SpawnableAssetRef>("SpawnableAssetRef")
+                ->Constructor()
+                ->Attribute(AZ::Script::Attributes::Scope, AZ::Script::Attributes::ScopeFlags::Common)
+                ->Attribute(AZ::Script::Attributes::Category, "Prefab/Spawning")
+                ->Attribute(AZ::Script::Attributes::Module, "prefabs")
+                ->Property("asset", BehaviorValueProperty(&SpawnableAssetRef::m_asset));
+        }
+    }
+
+    SpawnableAssetRef::SpawnableAssetRef()
+    {
+        OnSpawnAssetChanged();
+    }
+
+    SpawnableAssetRef::~SpawnableAssetRef()
+    {
+        AZ::Data::AssetBus::Handler::BusDisconnect();
+    }
+
+    SpawnableAssetRef::SpawnableAssetRef(const SpawnableAssetRef& rhs)
+        : m_asset(rhs.m_asset)
+    {
+        OnSpawnAssetChanged();
+    }
+
+    SpawnableAssetRef& SpawnableAssetRef::operator=(const SpawnableAssetRef& rhs)
+    {
+        m_asset = rhs.m_asset;
+        OnSpawnAssetChanged();
+        return *this;
+    }
+
+    void SpawnableAssetRef::OnSpawnAssetChanged()
+    {
+        // Disconnect from the bus beforehand in case the new asset is not valid
+        AZ::Data::AssetBus::Handler::BusDisconnect();
+
+        if (m_asset.GetId().IsValid())
+        {
+            AZStd::string rootSpawnableFile;
+            StringFunc::Path::GetFileName(m_asset.GetHint().c_str(), rootSpawnableFile);
+            StringFunc::Path::ReplaceExtension(rootSpawnableFile, Spawnable::DotFileExtension);
+            AZ::u32 rootSubId = SpawnableAssetHandler::BuildSubId(rootSpawnableFile);
```

Comment:
The usage of 'rootSpawnableFile' is confusing here because rootSpawnable refers to the spawnable generated from the root prefab in the editor. SpawnableAssetRef can hold an asset reference to any spawnable right? So this should be just called 'spwanableFile' or 'spawnablePath' ?